src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+
+        // Total count of minted token
+        pub TotalSupply get(total_supply): u64;
+
+        // Map of token index to token id
+        pub AllTokens get(token_by_index): map u64 => T::Hash;
+
+        // Last token index
+        pub AllTokensIndex: map T::Hash => u64;
+
+        // Map of account and index to token id
+        pub OwnedTokens get(token_of_owner_by_index): map (T::AccountId, u64) => T::Hash;
+
+        // Token id to index
+        pub OwnedTokensIndex: map T::Hash => u64;
+
+        // Value use to generate random value
+        pub Nonce: u64;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin  {
+        fn deposit_event() = default;
+
+        // Approve a token to an account
+        fn approve(origin, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let owner = match Self::owner_of(token_id) {
+                Some(c) => c,
+                None => return Err(Error::<T>::TokenNoOwner.into()),
+            };
+
+            ensure!(to != owner, "Owner is implicitly approved");
+            ensure!(sender == owner || Self::is_approved_for_all((owner.clone(), sender.clone())), "You are not allowed to approve for this token");
+
+            <TokenApprovals<T>>::insert(&token_id, &to);
+
+            Self::deposit_event(RawEvent::Approval(owner, to, token_id));
+
+            Ok(())
+        }
+
+        // Set if an operator can transfer an owner's token
+        fn set_approval_for_all(origin, to: T::AccountId, approved: bool) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            ensure!(to != sender, "You are already implicity approved for your own actions");
+            <OperatorApprovals<T>>::insert((sender.clone(), to.clone()), approved);
+
+            Self::deposit_event(RawEvent::ApprovalForAll(sender, to, approved));
+
+            Ok(())
+        }
+
+        // transfer_from will transfer to addresses even without a balance
+        // fn transfer_from(origin, from: T::AccountId, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+        //     let sender = ensure_signed(origin)?;
+        //     ensure!(Self::_is_approved_or_owner(sender, token_id), "You do not own this token");
+
+        //     Self::_transfer_from(from, to, token_id)?;
+
+        //     Ok(())
+        // }
+
+        // safe_transfer_from checks that the recieving address has enough balance to satisfy the ExistentialDeposit
+        // This is not quite what it does on Ethereum, but in the same spirit...
+        // fn safe_transfer_from(origin, from: T::AccountId, to: T::AccountId, token_id: T::Hash) -> DispatchResult {
+        //     let to_balance = <balances::Module<T>>::free_balance(&to);
+        //     ensure!(!to_balance.is_zero(), "'to' account does not satisfy the `ExistentialDeposit` requirement");
```

Comment:
I think the safe_transfer just do basic check, if there some deposit in account. In substreate, some currency needed to keep the account in balance pallet. I think the balance pallet need provide an interface to tell other runtime if an account is ready/valid to hold some assets.

Reply:
You are right about the balances pallet, but there is no requirement of a minimum balance to own an NFT.

Requiring a balance might indicate that "someone" is actually having a private key. 

The check in ERC 721 is different though, it check that the recipient is not a contract, which cannot send tokens anymore. Since we cannot do that check in substrate, I feel we should not expose a function with the same name, but completely different in behaviour.

My proposal is to remove `safe_transfer_from` and add a note to the README that explains why a check as in the ERC 721 specification is not possible in Substrate.

@manuel what are your thoughts about this?